OIDC: private key update (#9269)
mendix/docs: content/en/docs/appstore/use-content/platform-supported-content/modules/oidc.md

    For signing into the app, the OIDC SSO module will not use token introspection and will always validate against the published `jwks` endpoint.

* Stores an access token for each end-user that can be used to make API calls on their behalf

> * Can be configured to use either `client_secret_post`, `client_secret_basic`, or `private_key_jwt` as the client authentication method. Both `client_secret_post` and `client_secret_basic` make use of the `client-id` and `client-secret` as configured at the IdP. The `private_key_jwt` makes use of the `client-id`, **KeyPair Expiration Days**, and **JWT ALG(Signing Algorithm)** as configured at the IdP.

Reviewer: This is not configured "at the IdP"; it is configured "at the app", i.e. at the OIDC SSO module.
The OIDC module allows to be connected with multiple IdPs, so the configuration is at IdP-level

Author: done in the [commit](https://github.com/mendix/docs/pull/9269/commits/dabe9fdf68346e7ba009cb7d13743542d1c4034b)

    * OIDC SSO now supports Private Key JWT authentication, generating a key pair based on the selected signing algorithm and JWKS expiry duration. 

>     * It supports nine signing algorithms (ES256, ES384, ES512, PS256, PS384, PS512, RS256, RS384, RS512) and automatically regenerates a new key pair upon expiry.

Reviewer: This section is about protocol compliance and only for knowldegeable audience
": which parts of the spec do we support, which parts do we not support."
All the specifics of our implementation should go further down in the document.

So in adherence section only:
- Can be configured to use either `client_secret_post`, `client_secret_basic`, or `private_key_jwt` as the client authentication method.
- It supports nine signing algorithms (ES256, ES384, ES512, PS256, PS384, PS512, RS256, RS384, RS512)

The explanatiuon about key generation should be further down in the document, where we describe how to do configurations.

Author: Updated the Protocol Adherence section in this [commit](https://github.com/mendix/docs/pull/9269/commits/dabe9fdf68346e7ba009cb7d13743542d1c4034b). The deleted information has been included in the 'General OIDC Clients' section.

    For signing into the app, the OIDC SSO module will not use token introspection and will always validate against the published `jwks` endpoint.

* Stores an access token for each end-user that can be used to make API calls on their behalf

> * Can be configured to use either `client_secret_post`, `client_secret_basic`, or `private_key` as the client authentication method. Both `client_secret_post` and `client_secret_basic` make use of the `client-id` and `client-secret` as configured at the IdP. The `private_key` makes use of the `client-id`, **KeyPair Expiration Days**, and **JWT ALG(Signing Algorithm)** as configured at the IdP.

Reviewer: In specifications it's called private_key_jwt so we should call it like that and not another name, which causes confusion. Please update.

Author: updated in [this](https://github.com/mendix/docs/pull/9269/commits/6ba5d33c72079adf7383cb945a51353bd71e8701) commit

>     * OIDC SSO now supports Private Key JWT authentication, generating a key pair based on the selected signing algorithm and JWKS expiry duration. 

Reviewer: "now" should be in release notes, not in documentation.

Author: done in the [commit](https://github.com/mendix/docs/pull/9269/commits/dabe9fdf68346e7ba009cb7d13743542d1c4034b)

   **Client assertion** is automatically set to *Client ID and Secret*.

4. Choose the **Client authentication method** — make sure that you select a method that is supported by your IdP. You can normally check this via the `token_endpoint_auth_methods_supported` setting on the IdP’s well-known endpoint. Also, ensure that the correct client authentication method is configured at the IdP when you register the client.

    The options are:

>     * `client_secret_basic`: Your app will use the HTTP Basic Authentication scheme to authenticate itself at your IdP. (Default – for security reasons, this should be your preferred choice.) The `client_secret_basic` makes use of the `client-id` and `client-secret`.

Reviewer: Yes, it's okay to say that client_secret_basic is our default. But no longer for security reasons, so we need to remove that.

Reviewer 2: ```suggestion
    * `client_secret_basic`: Your app will use the HTTP Basic Authentication scheme to authenticate itself at your IdP. This is the default. The `client_secret_basic` makes use of the `client-id` and `client-secret`.
```

    * `client_secret_post`: Your app will authenticate itself by including its `client_id` and `client_secret` in the payload of token requests. (Older versions of the OIDC SSO module used this method.)

>     * `private_key_jwt`: This method uses asymmetric key cryptography (algorithm) for authentication. When you select the `private key` option, you can configure below fields:

Reviewer: FRom a security perspective private_key_jwt should now be the preferred option. Let's say that.

FYI: We may make this the default in the feature, but since this is a 'new' feature, it feels to soon to do that already now.

Reviewer 2: ```suggestion
    * `private_key_jwt`: This method uses asymmetric key cryptography (algorithm) for authentication. This is the best option for security. When you select the `private key` option, you can configure the following fields:
```

        * **Key Pair Expiration Days**: (default `90`)
        * **JWT ALG(Signing Algorithm)**: (default `RS256`)
 
    Once you **Save** the configuration, a key pair is automatically generated. Before you set up the private key

> authentication in your Mendix App, complete the JWKS configuration at your IdP, for example, Okta. For more information, see the [Configuring JWKS at Your IdP (Okta)](#jwks-okta) section. 

Reviewer: I'd like a bit less umphasis on Okta.

I suggest:
"..... Mendix App, complete the JWKS configuration at your IdP. Check the documentation of your IdP for details; if your using Okta, see the [Configuring JWKS at Your IdP (Okta)](#jwks-okta) section.

>     {{% alert color="info" %}} Requests signed with the new key may fail until Okta refreshes its key cache by calling the `/jwks` endpoint. {{% /alert %}}

Reviewer: Again let's make this a little bit more generic.

Suggestion: "Note that after a key renewal, some SSO requests may fail because your IdP may not immediately refresh its key cache."

Reviewer 2: ```suggestion
    {{% alert color="info" %}}After a key renewal, some SSO requests may fail if your IdP does not immediately refresh its key cache. {{% /alert %}}
```

4. Choose the **Client authentication method** — make sure that you select a method that is supported by your IdP. You can normally check this via the `token_endpoint_auth_methods_supported` setting on the IdP’s well-known endpoint. Also, ensure that the correct client authentication method is configured at the IdP when you register the client.

    The options are:
    * `client_secret_basic`: Your app will use the HTTP Basic Authentication scheme to authenticate itself at your IdP. (Default – for security reasons, this should be your preferred choice.) The `client_secret_basic` makes use of the `client-id` and `client-secret`.
    * `client_secret_post`: Your app will authenticate itself by including its `client_id` and `client_secret` in the payload of token requests. (Older versions of the OIDC SSO module used this method.)
    * `private_key_jwt`: This method uses asymmetric key cryptography (algorithm) for authentication. When you select the `private key` option, you can configure below fields:
        * **Key Pair Expiration Days**: (default `90`)
        * **JWT ALG(Signing Algorithm)**: (default `RS256`)
 

>     Once you **Save** the configuration, a key pair is automatically generated. Before you set up the private key
> authentication in your Mendix App, complete the JWKS configuration at your IdP, for example, Okta. For more information, see the [Configuring JWKS at Your IdP (Okta)](#jwks-okta) section. 

Reviewer 2: ```suggestion
    Once you **Save** the configuration, a key pair is automatically generated. Before you set up the private key authentication in your Mendix App, complete the JWKS configuration at your IdP. Check the documentation of your IdP for details. If you are using Okta, you can refer to the [Configuring JWKS at Your IdP (Okta)](#jwks-okta) section. 
```

    The options are:
    * `client_secret_basic`: Your app will use the HTTP Basic Authentication scheme to authenticate itself at your IdP. (Default – for security reasons this should be your preferred choice)
    * `client_secret_post`: Your app will authenticate itself by including its `client_id` and `client_secret` in the payload of token requests. (Older versions of the OIDC SSO module used this method).

>     * `private_key`: This method uses asymmetric key cryptography (algorithm) for authentication. When you select `private key` option, you can configure below fields:

Reviewer: should be "private_key_jwt". that's what I agreed with Phani.
was it not implemented this way?

Author: Already done in this [commit](https://github.com/mendix/docs/pull/9269/commits/6ba5d33c72079adf7383cb945a51353bd71e8701)

        * **Key Pair Expiration Days**: (default `90`)
        * **JWT ALG(Signing Algorithm)**: (default `RS256`)
 
    Once you **Save** the configuration, a key pair is automatically generated. Before you set up the private key
authentication in your Mendix App, complete the JWKS configuration at your IdP, for example, Okta. For more information, see the [Configuring JWKS at Your IdP (Okta)](#jwks-okta) section. 

>     {{% alert color="warning" %}} The JWKS expires in one minute, which results in the deletion of all previous

Reviewer: No, the JWKS expured after 90 days. Also the "Ensure..." does not make sense; the customer doesn't configure any such thing at all.

The assertion in the request is valid for one minute.
I suggest we do NOT talk about this at all in the diocumentation.

I suggest to remove all of this:
"The JWKS expires....... autrhentication failures."

Author: done in the [commit](https://github.com/mendix/docs/pull/9269/commits/dabe9fdf68346e7ba009cb7d13743542d1c4034b)

* **ClientAuthenticationMethod** (*default: client_secret_basic*) – the client authentication method — the caption of OIDC.ENU_ClientAuthenticationMethod

>     Examples: `client_secret_post`, `client_secret_basic`, or `private_key`

Reviewer: private_key_jwt

Author: Already done in this [commit](https://github.com/mendix/docs/pull/9269/commits/6ba5d33c72079adf7383cb945a51353bd71e8701).

{{% alert color="info" %}}

> when you set **ClientAuthenticationMethod** as `private_key`, you do not need to set **ClientSecret** constant.

Reviewer: private_key_jwt

Author: Already changed in this [commit](https://github.com/mendix/docs/pull/9269/commits/6ba5d33c72079adf7383cb945a51353bd71e8701)
